### PR TITLE
SWDEV-423857

### PR DESCRIPTION
**Pull request by YvanMokwinski**
SWDEV-423857

- proper error message for failure due to missing files
- ship cmake script and executable to download and convert matrix files

### CODE REVIEW

File: clients/include/utility.hpp [hunk 1]
```diff
@@ -6043,6 +6043,35 @@ class Arguments
...
+inline void missing_file_error_message(const char* filename)
+{
+    std::cerr << "#" << std::endl;
+    std::cerr << "# error:" << std::endl;
+    std::cerr << "# cannot open file '" << filename << "'" << std::endl;
+    std::cerr << "#" << std::endl;
+    std::cerr << "# PLEASE READ CAREFULLY !" << std::endl;
+    std::cerr << "#" << std::endl;
+    std::cerr << "# What could be the reason of this error: " << std::endl;
+    std::cerr << "# You are running the testing application then it expects to find the file "
```
Comment (jsandham) [inline clients/include/utility.hpp:6055]: application and it expects

or you did not specify the location

Comment (jsandham) [reply clients/include/utility.hpp:6055]: matrices, or you did not

File: clients/include/utility.hpp [hunk 2]
```diff
@@ -6043,6 +6043,35 @@ class Arguments
...
+                 "files. If you want to specify the location of the folder containing your files "
+                 "then you will find the needed information with 'hipsparse-test --help'."
+                 "If you need to download matrices, a cmake script "
+                 "'hipsparse_clientmatrices.cmake' is available from the hipsparse client package."
+              << std::endl;
+    std::cerr << "#" << std::endl;
+    std::cerr
+        << "# Examples: 'hipsparse_clientmatrices.cmake -DCMAKE_MATRICES_DIR=<path-of-your-folder>'"
+        << std::endl;
+    std::cerr << "#           'hipsparse-test --matrices-dir <path-of-your-folder>'" << std::endl;
```
Comment (jsandham) [inline clients/include/utility.hpp:6067]: Just double checking, does hipsparse also have the --matrices-dir option?

File: cmake/hipsparse_clientmatrices.cmake [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  list(GET TEST_MATRICES ${i} m)
+  list(GET TEST_MD5HASH ${i} md5)
+
+  string(REPLACE "/" ";" sep_m ${m})
+  list(GET sep_m 0 dir)
+  list(GET sep_m 1 mat)
+
+  # Download test matrices if not already downloaded
+  if(NOT EXISTS "${CMAKE_MATRICES_DIR}/${mat}.bin")
+    if(NOT ROCSPARSE_MTX_DIR)
```
Comment (jsandham) [inline cmake/hipsparse_clientmatrices.cmake:96]: Are these references to ROCSPARSE correct?

Comment (YvanMokwinski) [reply cmake/hipsparse_clientmatrices.cmake:96]: good catch

**Review by jsandham — APPROVED** [review]

File: cmake/hipsparse_clientmatrices.cmake (hunk 3 above)
Comment (YvanMokwinski) [reply cmake/hipsparse_clientmatrices.cmake:96]: it's actually an existing glitch from ClientMatrices.cmake

File: clients/include/utility.hpp (hunk 2 above)
Comment (YvanMokwinski) [reply clients/include/utility.hpp:6067]: check hipsparse-test --help

File: clients/include/utility.hpp (hunk 1 above)
Comment (jsandham) [reply clients/include/utility.hpp:6055]: Sorry that first one should be:

...application, then it expects...

File: cmake/hipsparse_clientmatrices.cmake (hunk 3 above)
Comment (YvanMokwinski) [reply cmake/hipsparse_clientmatrices.cmake:96]: I don't know what this ROCSPARS_MTX_DIR is doing anyway, I should ask @ntrost57

File: clients/include/utility.hpp (hunk 1 above)
Comment (YvanMokwinski) [reply clients/include/utility.hpp]: "You are running the testing application and it expects"

**Review by ntrost57 — APPROVED** [review]

**Pull request merged**